### ISSUE TITLE
[NSM] NSM should not auto-save when switching sessions or closing

### ISSUE THREAD

**Issue by ViktorNova** [original-male/non#149]
I know this is a bold statement, so please hear me out - 

NSM's function of automatically saving the session when switching sessions or closing seems very counter-intuitive to the expected behavior of the user interface, and also is in conflict with how most software generally works, since users have come to expect that unless we explicitly tell our software to save, it will not decide to save on it's own  

The reason I think this is a big problem for NSM users is that this behavior makes the bold assumption that any and all  change made to all session clients and the current snapshot of all audio/midi routing systemwide (if jackpatch is part of the session) should be committed to permanent memory, overwriting any previous setup by default, any time a session is switched or closed, without the user taking any action that explicitly tells it to do this. Combined with the presence of a master save button and the save button on each client, NSM really makes it seem like 'it won't save unless I tell it to'. The only way around this is to **abort** the current session, and confirm in the dialog box that you want to discard any unsaved changes. 

What I'm proposing is that it should be the other way around - sessions should switch **without** saving by default, and the **Close** button should be renamed to **Save & close**. Not only would this make the functionality more transparent, but it would make it easier to save specific clients only without then having to abort the session before switching to a new one. I also think that if this became the new behavior, that NSM should pop up a confirm dialog when you switch sessions, warning you that unsaved changes will be lost. I realize NSM can't be aware of whether or not non-NSM clients have unsaved changes, but even so, having this dialog pop up on every switch regardless would be a nice thing anyhow, in case you accidentally click/tap a different session while in the middle of performing a song (I've done it!)

To put this in context, here's some personal examples off the top of my head of how the current behavior bites me (i use NSM for sequencing MIDI/DMX/video live & at band practice)
   - If I'm in the middle of playing a set and forget to hit abort before switching to the next song, the next time I play the set the song position will start at the end of the song..awkward!
   - If I'm working out beats on the road and the laptop is not plugged into all the external MIDI gear, all those connections will be lost next time if I forget to stop the jackpatch client before switching songs

Wow I just typed a novel.. =) 
What do you think about all this?

**Comment by ViktorNova** [comment]
I have a question about this actually, is the auto saving behavior part of nsmd, or is it only the way the GUI client works?

**Comment by nico202** [comment]
I found this behavior to be life-saving sometimes, but I understand your
problem.. Maybe a session - dependent switch?
[  ] autosave

Il giorno sab 10 gen 2015 23:18 Viktor Nova <notifications@github.com> ha
scritto:

**Comment by original-male** [comment]
@ViktorNova The functionality is built into NSMD. That is, the response to the 'open' message is to attempt to save the active session before opening the new one. Not that that makes it impossible to change the behavior...

It's a tricky issue. Regarding your example scenarios--there's no policy anywhere about whether JACK clients should save/restore the transport position. Non-Timeline throws the further wrench into the situation in that it doesn't have a 'save' at all--every action is saved the moment it happens via the journal, so even if you Abort a session, there's no guarantee that some clients won't have saved some of your changes. Again, nowhere is there any policy for this kind of thing. In your example where JACKPatch forgets connections to/from missing devices... Well, that might be better solved by adding some functionality to JACKPatch (and JACK and patchage or whatever).  A piece of the conncetion management puzzle that's always been missing is that there's no difference, as far as the JACK API is concerned, between a connection being lost/dropped and a connection being explicitly broken by the user.

**Comment by ViktorNova** [comment]
@original-male Right, yeah that makes sense - it's definitely not a perfect landscape any way you look at it, although NSM is already pretty key in making it all work together, it's closer than it's ever been! I suppose to make it work the best for everyone, it would be ideal to have this behavior configurable, either for NSMD globally, per session, or per-client.. perhaps an NSMD directive in addition to 'open' such as 'open-dontsave'. How do you feel about this?

To clarify - in my example of the transport position thing, the reason it is a problem when NSM is used live is because when the session is switched without aborting, a save signal is sent to the client who is transport master (in my case [los](https://github.com/falkTX/los) ) which saves the song position at the end of the song before switching to the new session, so the next time the song is opened, it starts at the end

For persistent JACK/ALSA connections, I'm working right now on a fork of aj-snapshot that makes it NSM aware with the ability to save the patchbay state outside of NSM (and ignore NSM save requests) and hope to have that code up soon!

**Comment by chipship** [comment]
@ViktorNova although I haven't had a chance to use NSM live yet, but from my experience on previous live performances using other tools (FL, Ableton) I noticed that I tend to ruin current save sessions in the heat of the moment even without an autosave.
So I developed a habit of duplicating my sessions, loading a temporary one for the performance. When I'm back home I might review the saves from the performance and export anything worthwhile.

I agree the current way is confusing a bit, and should be made clearer on the GUI, but once you know how it works you can use it for your advantage.
My intention is to have sessions I can bounce back and forth between live, so whenever I bonce back to a previous session it's loaded into its last state.

Regardless, it's generally a good practice to not directly work on any saves you actually want to save.
